### PR TITLE
Work around compilation events appearing in the queue too late.

### LINKED ISSUE #8178: Crash in diagnostics analyzer
Was editing in roslyn when i got this:

```
Framework Version: v4.0.30319
Description: The application requested process termination through System.Environment.FailFast(string message).
Message: System.InvalidOperationException: Cannot call Enqueue when the queue is already completed.
   at Microsoft.CodeAnalysis.Diagnostics.AsyncQueue`1.Enqueue(TElement value)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.SymbolDeclaredEvent(Symbol symbol)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.GetNameToMembersMap()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.GetMembers(String name)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredMember(NamespaceOrTypeSymbol container, TextSpan declarationSpan, String name)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredNamedType(CSharpSyntaxNode declarationSyntax, String name)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredType(BaseTypeDeclarationSyntax declarationSyntax)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredSymbol(BaseTypeDeclarationSyntax declarationSyntax, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetDeclaredSymbolCore(SyntaxNode declaration, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.DeclarationComputer.GetDeclaredSymbol(SemanticModel model, SyntaxNode node, Boolean getSymbol, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.DeclarationComputer.GetDeclarationInfo(SemanticModel model, SyntaxNode node, Boolean getSymbol, IEnumerable`1 executableCodeBlocks, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.DeclarationComputer.GetDeclarationInfo(SemanticModel model, SyntaxNode node, Boolean getSymbol, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpDeclarationComputer.ComputeDeclarations(SemanticModel model, SyntaxNode node, Func`3 shouldSkip, Boolean getSymbol, List`1 builder, Nullable`1 levelsToCompute, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpDeclarationComputer.ComputeDeclarations(SemanticModel model, SyntaxNode node, Func`3 shouldSkip, Boolean getSymbol, List`1 builder, Nullable`1 levelsToCompute, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpDeclarationComputer.ComputeDeclarationsInNode(SemanticModel model, SyntaxNode node, Boolean getSymbol, List`1 builder, CancellationToken cancellationToken, Nullable`1 levelsToCompute)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.ComputeDeclarationsInNode(SyntaxNode node, Boolean getSymbol, List`1 builder, CancellationToken cancellationToken, Nullable`1 levelsToCompute)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ComputeDeclarationAnalysisData(ISymbol symbol, SyntaxReference declaration, SemanticModel semanticModel, Boolean shouldExecuteSyntaxNodeActions, AnalysisScope analysisScope, Func`1 allocateData, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.<>c__DisplayClass37_0.<ExecuteDeclaringReferenceActions>b__0(Func`1 allocateData)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.CompilationData.GetOrComputeDeclarationAnalysisData(SyntaxReference declaration, Func`2 computeDeclarationAnalysisData, Boolean cacheAnalysisData)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ExecuteDeclaringReferenceActions(SyntaxReference decl, Int32 declarationIndex, SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisStateOpt, Boolean shouldExecuteSyntaxNodeActions, Boolean shouldExecuteOperationActions, Boolean shouldExecuteCodeBlockActions, Boolean shouldExecuteOperationBlockActions, Boolean isInGeneratedCode, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1.ExecuteDeclaringReferenceActions(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisStateOpt, Boolean isGeneratedCodeSymbol, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessSymbolDeclared(SymbolDeclaredCompilationEvent symbolEvent, AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessEventCore(CompilationEvent e, AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.<ProcessCompilationEventsCoreAsync>d__64.MoveNext()
Stack:
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.CodeAnalysis.FatalError.Report(System.Exception, System.Action`1<System.Exception>)
   at Microsoft.CodeAnalysis.FatalError.ReportUnlessCanceled(System.Exception)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+<ProcessCompilationEventsCoreAsync>d__64.MoveNext()
   at Microsoft.CodeAnalysis.Diagnostics.AsyncQueue`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Enqueue(System.__Canon)
   at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.SymbolDeclaredEvent(Microsoft.CodeAnalysis.CSharp.Symbol)
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.GetNameToMembersMap()
   at Microsoft.CodeAnalysis.CSharp.Symbols.SourceNamespaceSymbol.GetMembers(System.String)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredMember(Microsoft.CodeAnalysis.CSharp.Symbols.NamespaceOrTypeSymbol, Microsoft.CodeAnalysis.Text.TextSpan, System.String)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredNamedType(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxNode, System.String)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredType(Microsoft.CodeAnalysis.CSharp.Syntax.BaseTypeDeclarationSyntax)
   at Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDeclaredSymbol(Microsoft.CodeAnalysis.CSharp.Syntax.BaseTypeDeclarationSyntax, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.GetDeclaredSymbolCore(Microsoft.CodeAnalysis.SyntaxNode, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.DeclarationComputer.GetDeclaredSymbol(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.SyntaxNode, Boolean, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.DeclarationComputer.GetDeclarationInfo(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.SyntaxNode, Boolean, System.Collections.Generic.IEnumerable`1<Microsoft.CodeAnalysis.SyntaxNode>, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.DeclarationComputer.GetDeclarationInfo(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.SyntaxNode, Boolean, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpDeclarationComputer.ComputeDeclarations(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.SyntaxNode, System.Func`3<Microsoft.CodeAnalysis.SyntaxNode,System.Nullable`1<Int32>,Boolean>, Boolean, System.Collections.Generic.List`1<Microsoft.CodeAnalysis.DeclarationInfo>, System.Nullable`1<Int32>, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpDeclarationComputer.ComputeDeclarations(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.SyntaxNode, System.Func`3<Microsoft.CodeAnalysis.SyntaxNode,System.Nullable`1<Int32>,Boolean>, Boolean, System.Collections.Generic.List`1<Microsoft.CodeAnalysis.DeclarationInfo>, System.Nullable`1<Int32>, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.CSharp.CSharpDeclarationComputer.ComputeDeclarationsInNode(Microsoft.CodeAnalysis.SemanticModel, Microsoft.CodeAnalysis.SyntaxNode, Boolean, System.Collections.Generic.List`1<Microsoft.CodeAnalysis.DeclarationInfo>, System.Threading.CancellationToken, System.Nullable`1<Int32>)
   at Microsoft.CodeAnalysis.CSharp.CSharpSemanticModel.ComputeDeclarationsInNode(Microsoft.CodeAnalysis.SyntaxNode, Boolean, System.Collections.Generic.List`1<Microsoft.CodeAnalysis.DeclarationInfo>, System.Threading.CancellationToken, System.Nullable`1<Int32>)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1[[Microsoft.CodeAnalysis.CSharp.SyntaxKind, Microsoft.CodeAnalysis.CSharp, Version=1.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]].ComputeDeclarationAnalysisData(Microsoft.CodeAnalysis.ISymbol, Microsoft.CodeAnalysis.SyntaxReference, Microsoft.CodeAnalysis.SemanticModel, Boolean, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, System.Func`1<DeclarationAnalysisData>, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1+<>c__DisplayClass37_0[[Microsoft.CodeAnalysis.CSharp.SyntaxKind, Microsoft.CodeAnalysis.CSharp, Version=1.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]].<ExecuteDeclaringReferenceActions>b__0(System.Func`1<DeclarationAnalysisData>)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+CompilationData.GetOrComputeDeclarationAnalysisData(Microsoft.CodeAnalysis.SyntaxReference, System.Func`2<System.Func`1<DeclarationAnalysisData>,DeclarationAnalysisData>, Boolean)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1[[Microsoft.CodeAnalysis.CSharp.SyntaxKind, Microsoft.CodeAnalysis.CSharp, Version=1.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]].ExecuteDeclaringReferenceActions(Microsoft.CodeAnalysis.SyntaxReference, Int32, Microsoft.CodeAnalysis.Diagnostics.SymbolDeclaredCompilationEvent, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalysisState, Boolean, Boolean, Boolean, Boolean, Boolean, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver`1[[Microsoft.CodeAnalysis.CSharp.SyntaxKind, Microsoft.CodeAnalysis.CSharp, Version=1.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]].ExecuteDeclaringReferenceActions(Microsoft.CodeAnalysis.Diagnostics.SymbolDeclaredCompilationEvent, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalysisState, Boolean, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessSymbolDeclared(Microsoft.CodeAnalysis.Diagnostics.SymbolDeclaredCompilationEvent, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalysisState, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessEventCore(Microsoft.CodeAnalysis.Diagnostics.CompilationEvent, Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalysisState, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+<ProcessCompilationEventsCoreAsync>d__64.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Start[[Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+<ProcessCompilationEventsCoreAsync>d__64, Microsoft.CodeAnalysis, Version=1.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<ProcessCompilationEventsCoreAsync>d__64 ByRef)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver.ProcessCompilationEventsCoreAsync(Microsoft.CodeAnalysis.Diagnostics.AnalysisScope, Microsoft.CodeAnalysis.Diagnostics.AnalysisState, Boolean, System.Threading.CancellationToken)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+<>c__DisplayClass63_0+<<ProcessCompilationEventsAsync>b__0>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].Start[[Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+<>c__DisplayClass63_0+<<ProcessCompilationEventsAsync>b__0>d, Microsoft.CodeAnalysis, Version=1.2.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]](<<ProcessCompilationEventsAsync>b__0>d ByRef)
   at Microsoft.CodeAnalysis.Diagnostics.AnalyzerDriver+<>c__DisplayClass63_0.<ProcessCompilationEventsAsync>b__0()
   at System.Threading.Tasks.Task`1[[System.__Canon, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

### PR DESCRIPTION
**Pull request by gafter**
I suspect this occurs due to syntax error recovery because the semantic model does not
find appropriate symbols in the symbol table, and creates them as needed, but I was unable to reproduce the issue to verify that. We prevent the crash by suppressing events that arise too late.
Fixes #8178 

@dotnet/roslyn-compiler Please review.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -2885,7 +2885,10 @@ internal override AnalyzerDriver AnalyzerForLanguage(ImmutableArray<DiagnosticAn
 
         internal void SymbolDeclaredEvent(Symbol symbol)
         {
-            if (EventQueue != null) EventQueue.Enqueue(new SymbolDeclaredCompilationEvent(this, symbol));
+            if (EventQueue?.IsCompleted == false)
+            {
+                EventQueue.Enqueue(new SymbolDeclaredCompilationEvent(this, symbol));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Can we assert that this is in fact an error case? For example, we can verify that the symbol is not reachable by descending from source module's global namespace.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Perhaps it is even a good idea to throw if the symbol is reachable.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: Perhaps we shouldn't raise events for unreachable symbols even when queue is still "open".

**Comment by AlekseyTs** [comment]
Have you tried reproducing the problem by declaring an event within a namespace?

**Comment by cston** [comment]
Could a similar issue exist in VB?

**Comment by cston** [comment]
`AnalyzerDriver.AttachQueueAndStartProcessingEvents` will call `EventQueue.TryComplete` when the `CancellationToken` is cancelled. Is there a potential race between that call to `TryComplete` and any events being generated within the compiler? If so, perhaps all calls to `EventQueue.Enqueue` in the compiler should use `TryEnqueue` instead.

**Comment by gafter** [comment]
> AnalyzerDriver.AttachQueueAndStartProcessingEvents will call EventQueue.TryComplete when the CancellationToken is cancelled. Is there a potential race between that call to TryComplete and any events being generated within the compiler? If so, perhaps all calls to EventQueue.Enqueue in the compiler should use TryEnqueue instead.

Aha! That could cause this crash! I will rework to use TryEnqueue.

**Comment by jaredpar** [comment]
Good observation @cston!

> Aha! That could cause this crash! I will rework to use TryEnqueue.

That seems like it would take away some validation we have today.  There seem to be two reasons that `Enqueue` could fail:

1. A use of `Enqueue` after cancellation occurred (scenario called out by @cston)
2. A use of `Enqueue` after the queue was completed.  

The first represents a correctly functioning program while the second represents a bug in the compiler.  Won't switching all calls to `TryEnqueue` normalize the two cases and reduce our ability to detect the second one?  

Perhaps we should add a new API: `EnqueueUnlessCancelled` that the compiler could use?  Or simply change `Enqueue` to not throw if cancellation occurred.

**Comment by gafter** [comment]
Closing to replace with a *correct* fix including test on the stabilization branch.

**Pull request closed without merging**